auth/signaturevalidator.js

```diff
@@ -71,23 +85,48 @@ var limitClientWithExt = function(client, ext, expandScopes) {
       throw new Error("ext.certificate cannot last longer than 31 days!");
     }
 
-    // Check scope validity
+    // Check clientId validity
+    if (issuingClientId !== credentialName) {
+      if (!cert.clientId) {
+        throw new Error('ext.certificate.clientId must be set when ext.certificate.issuer is given');
+      }
+      if (typeof(cert.clientId) !== 'string') {
+        throw new Error('ext.certificate.clientId must be a string');
+      }
+      if (cert.clientId !== credentialName) {
+        throw new Error('ext.certificate.clientId must match the supplied clientId');
+      }
+      let createScope = 'auth:create-client:' + credentialName;
+      if (!utils.scopeMatch(issuingScopes, [[createScope]])) {
+        throw new Error("ext.certificate issuer `" + issuingClientId +
+                        "` doesn't have `" + createScope + "` for supplied clientId.");
+      }
+    } else {
+      if (cert.hasOwnProperty('clientId')) {
+        throw new Error('ext.certificate.clientId must only be used with ext.certificate.issuer');
+      }
+    }
 
     // Validate certificate scopes are subset of client
-    if (!utils.scopeMatch(client.scopes, [cert.scopes])) {
-      throw new Error("ext.certificate issuer `" + client.clientId +
+    if (!utils.scopeMatch(scopes, [cert.scopes])) {
+      throw new Error("ext.certificate issuer `" + issuingClientId +
                       "` doesn't have sufficient scopes");
     }
 
     // Generate certificate signature
-    var signature = crypto.createHmac('sha256', client.accessToken)
-      .update([
-        'version:'  + '1',
-        'seed:'     + cert.seed,
-        'start:'    + cert.start,
-        'expiry:'   + cert.expiry,
-        'scopes:',
-      ].concat(cert.scopes).join('\n'))
+    var sigContent = []
+    sigContent.push('version:'    + '1');
+    if (cert.clientId && cert.issuer) {
+      sigContent.push('clientId:' + cert.clientId);
```

Comment:
Why would certificate include `clientId`? Let's not duplicate...

Just use `credentialName` instead of `cert.clientId`